core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/iterator/PathIteration.java

```diff
@@ -190,6 +195,16 @@ protected BindingSet getNextElement() throws QueryEvaluationException {
 		}
 	}
 
+	private void addBinding(BindingSet bs, String name, Value value) {
+		if (bs instanceof QueryBindingSet) {
+			((QueryBindingSet) bs).addBinding(name, value);
+		} else if (bs instanceof MapBindingSet) {
+			((MapBindingSet) bs).addBinding(name, value);
+		} else {
+			throw new IllegalStateException("Unexpected BindingSet implementation: " + bs.getClass());
```

Comment:
FWIW I have long thought that we are using the various bindingset implementations inconsistently throughout the framework. It could do with straightening out.